presto-main/src/main/java/com/facebook/presto/Session.java

```diff
@@ -131,7 +133,8 @@ public Session(
             Map<SqlFunctionId, SqlInvokedFunction> sessionFunctions,
             Optional<Tracer> tracer,
             WarningCollector warningCollector,
-            RuntimeStats runtimeStats)
+            RuntimeStats runtimeStats,
+            Optional<QueryType> queryType)
```

Comment:
why is this optional ? can a session have this empty ?

Reply:
I don't believe there is an instance where queryType can be empty. If my understanding is correct, all queries that make it to DispatchManager will have some enum of QueryType. However, as a precaution, I wrapped QueryType in an optional.

Reply:
so then lets have it as NOT optional. And fix any issues that my arise.

Reply:
After making the changes, I realized that there are some areas where I put an empty optional instead of null.
Example would be here: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java#L308

Since the [session ctor](https://github.com/prestodb/presto/blob/33ff9521dbd25e2b7b1181cf01391cb899857d65/presto-main/src/main/java/com/facebook/presto/Session.java#L109) requires all fields to be non-null, I believe it may be the design appropriate choice to keep the queryType as an Optional.

Reply:
Why are we not adding QueryType to SessionRepresentation as well ? Then we do need NOT send Optional.empty() or null right ?

Reply:
After making the change in SessionRepresentation and other places, there is another issue if we don't use an Optional. queryType is not defined when we build session. We pass in session in a few [places](https://github.com/prestodb/presto/blob/2f4ea64d3bf3fa8c89b3e805d5d0987ac6d971c2/presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchManager.java#L274) before we even have queryType. So if we want to maintain the non-null rule in session ctor, I believe using an Optional is the best bet.

Reply:
It is a little confusing but I am more than happy to discuss the thought process!

Reply:
@NikhilCollooru the challenge here is the Session usage in DispatchManager#createQueryInternal:
1. Session is instantiated and passed to createAnalyzerOptions()
2. Same session is passed to getAnalyzerType()
3. The SQL string is only parsed on the following line in QueryPreparer#prepareQuery

So currently session is being used before we know what the QueryType actually is.  Now the Session usage in (1) and (2) is only to extract system property values, so we could try to separate that concern from the Session representation itself, pulling out SystemPropertyManager and the Map<String, String> of system properties.  But that looks like a big refactoring as the use of Session as an interface to system properties seems pretty deeply embedded.

Any thoughts on how we can better clean this up?  The current use of Session in DispatchManager#createQueryInternal is a bit wonky to begin with as Session is first created from SessionSupplier (not the builder for some reason), then recreated by SessionPropertyDefaults#newSessionQithDefaultProperties.  I don't really understand why the SessionPropertyDefaults part happens _after_ the analyzer and query preparer setup.